### PR TITLE
SONARJNKNS-253 Make Inject SonarQube environment (SonarBuildWrapper) …

### PR DESCRIPTION
**Pull request by henryju**
…compatible with pipeline

### CODE REVIEW

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. CRITICAL [SonarInstallation.java#L308](https://github.com/SonarSource/sonar-scanner-jenkins/blob/faa9d70e0e6a6d680008617fd1e1fa5de8e72a21/src/main/java/hudson/plugins/sonar/SonarInstallation.java#L308): Make this class "private" or elevate the visibility of "readResolve". [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2062)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. CRITICAL [SonarInstallation.java#L308](https://github.com/SonarSource/sonar-scanner-jenkins/blob/f6e717b56b7e46f63ef72e7357530de4258a99f7/src/main/java/hudson/plugins/sonar/SonarInstallation.java#L308): Make this class "private" or elevate the visibility of "readResolve". [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2062)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. CRITICAL [SonarInstallation.java#L308](https://github.com/SonarSource/sonar-scanner-jenkins/blob/addff46bb23e951f1b5113e9bc5a3b9d63252b60/src/main/java/hudson/plugins/sonar/SonarInstallation.java#L308): Make this class "private" or elevate the visibility of "readResolve". [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2062)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. CRITICAL [SonarInstallation.java#L308](https://github.com/SonarSource/sonar-scanner-jenkins/blob/7b86bfe1e7f2c0229000c920845d1dbab3e2638a/src/main/java/hudson/plugins/sonar/SonarInstallation.java#L308): Make this class "private" or elevate the visibility of "readResolve". [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2062)

File: src/main/java/hudson/plugins/sonar/SonarBuildWrapper.java [hunk 1]
```diff
@@ -97,12 +176,47 @@ public OutputStream decorateLogger(AbstractBuild build, OutputStream outputStrea
...
+    }
+
+    @Override
+    public void tearDown(Run<?, ?> build, FilePath workspace, Launcher launcher, TaskListener listener) throws IOException, InterruptedException {
+      // null result means success so far. If no logs are found, it's probably because it was simply skipped
+      if (build.getResult() == null) {
+        SonarUtils.addBuildInfoTo(build, installation.getName(), true);
+      } else {
+        SonarUtils.addBuildInfoTo(build, installation.getName(), false);
+      }
```
Comment (janos-ss) [inline src/main/java/hudson/plugins/sonar/SonarBuildWrapper.java]: How about putting the condition in the 3rd param to avoid duplication?

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. CRITICAL [SonarInstallation.java#L308](https://github.com/SonarSource/sonar-scanner-jenkins/blob/d74aa1bdcac99bd00e95b33f3677b4757da26780/src/main/java/hudson/plugins/sonar/SonarInstallation.java#L308): Make this class "private" or elevate the visibility of "readResolve". [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2062)

File: src/main/java/hudson/plugins/sonar/SonarBuildWrapper.java [hunk 2]
```diff
@@ -72,15 +75,91 @@ public SonarBuildWrapper(@Nullable String installationName) {
...
+    return sonarEnvResolved;
+  }
+
+  private static String getAdditionalProps(SonarInstallation inst) {
+    ArgumentListBuilder builder = new ArgumentListBuilder();
+    // no need to tokenize since we need a String anyway
+    builder.add(inst.getAdditionalAnalysisPropertiesUnix());
+    builder.add(inst.getAdditionalProperties());
+
+    return StringUtils.join(builder.toList(), ' ');
```
Comment (janos-ss) [inline src/main/java/hudson/plugins/sonar/SonarBuildWrapper.java]: How about using `String.join` instead?

File: src/main/java/hudson/plugins/sonar/SonarBuildWrapper.java [hunk 3]
```diff
@@ -72,15 +75,91 @@ public SonarBuildWrapper(@Nullable String installationName) {
...
+    }
+    sb.append("}");
+
+    map.put("SONARQUBE_SCANNER_PARAMS", sb.toString());
+
+    // resolve variables against each other
+    Map<String, String> sonarEnvResolved = new HashMap<>(map);
+    EnvVars.resolve(sonarEnvResolved);
+
+    return sonarEnvResolved;
```
Comment (janos-ss) [inline src/main/java/hudson/plugins/sonar/SonarBuildWrapper.java]: Just out of curiosity, why do we need a new map here? Can't you just `EnvVars.resolve(map)` and `return map`?

**Comment by janos-ss** [comment]
Done reviewing, see my comments and QG, LGTM.

File: src/main/java/hudson/plugins/sonar/SonarBuildWrapper.java (hunk 1 above)
Comment (henryju) [inline src/main/java/hudson/plugins/sonar/SonarBuildWrapper.java]: Indeed. that should even deserve a rule in SonarQube :)

File: src/main/java/hudson/plugins/sonar/SonarBuildWrapper.java (hunk 2 above)
Comment (henryju) [inline src/main/java/hudson/plugins/sonar/SonarBuildWrapper.java]: Not sure it behave exactly the same regarding null / empty strings. So I'll not risk a regression :) BTW it is not new code (just moved).

File: src/main/java/hudson/plugins/sonar/SonarBuildWrapper.java (hunk 3 above)
Comment (henryju) [inline src/main/java/hudson/plugins/sonar/SonarBuildWrapper.java]: Good catch, I don't think there is a reason.

**Pull request merged**